### PR TITLE
Move correct SoR index discovery to Backfiller resource

### PR DESCRIPTION
**Pull request by sisyphusSmiling**
Related: #414
Targets: #416

### CODE REVIEW

File: contracts/RandomBeaconHistory.cdc [hunk 1]
```diff
@@ -126,32 +146,33 @@ access(all) contract RandomBeaconHistory {
...
-            var count = UInt64(0)
+            var count = 0 as UInt64
             while count < self.maxEntriesPerCall {
                 // move to the next empty entry
                 while index < arrayLength && RandomBeaconHistory.randomSourceHistory[index] != [] {
                     index = index + 1
                 }
                 // if we reach the end of the array then all existing gaps got filled
-                if index == arrayLength {
+                if index >= arrayLength {
```
Comment (tarakby) [inline contracts/RandomBeaconHistory.cdc:175]: If we keep the `return` above, then this change isn't needed.

File: contracts/RandomBeaconHistory.cdc [hunk 2]
```diff
@@ -126,32 +146,33 @@ access(all) contract RandomBeaconHistory {
...
+        access(contract) fun backfill(randomSource: [UInt8]) {
+
+            let correctIndex = self.findGapAndReturnCorrectIndex()
+            var arrayLength = UInt64(RandomBeaconHistory.randomSourceHistory.length)
             // optional optimization for the happy common path: if no gaps are detected, 
             // backfilling isn't needed, return early
             if correctIndex == self.gapStartIndex  {
-                self.gapStartIndex = currentArrayLength + 1
-                return
+                self.gapStartIndex = arrayLength + 1
```
[the remark this answers is not in the document]
Comment (sisyphusSmiling) [reply contracts/RandomBeaconHistory.cdc:156]: My mistake, I'll push a change with the early return

[the remark this answers is not in the document]
Comment (sisyphusSmiling) [reply contracts/RandomBeaconHistory.cdc:156]: Corrected in [c3d8cca](https://github.com/onflow/flow-core-contracts/pull/418/commits/c3d8cca24884d6d8a752fe93d13ad816863c085b)

**Review by tarakby — APPROVED** [review]
Sweet! ready to merge now 👌🏼

**Pull request merged**